test/test-pipe-try-write.c

```diff
@@ -0,0 +1,125 @@
+#include "uv.h"
+#include "task.h"
+
+static void (*spam)(uv_pipe_t* handle);
+static uv_pipe_t client_handle;
+static uv_pipe_t peer_handle;
+static uv_pipe_t server_handle;
+static uv_write_t write_req;
+
+
+static void write_cb(uv_write_t* req, int status) {
+  ASSERT_EQ(0, status);
+}
+
+
+static void spam_0(uv_pipe_t* handle) {
+  uv_buf_t buf;
+
+  buf = uv_buf_init("", 0);
+  ASSERT_EQ(0, uv_try_write((uv_stream_t*) handle, &buf, 1));
+
+  /* Non-empty write to start the event loop moving. */
+  buf = uv_buf_init("hello, world", sizeof("hello, world") - 1);
+  ASSERT_EQ(0, uv_write(&write_req, (uv_stream_t*) handle, &buf, 1, write_cb));
+}
+
+
+static void spam_1(uv_pipe_t* handle) {
+  uv_buf_t buf;
+  int rc;
+
+  buf = uv_buf_init("hello, world", sizeof("hello, world") - 1);
+  do
+    rc = uv_try_write((uv_stream_t*)handle, &buf, 1);
+  while (rc > 0);
+
+  ASSERT_EQ(rc, UV_EAGAIN);
+}
+
+
+static void spam_2(uv_pipe_t* handle) {
+  uv_buf_t bufs[2];
+  int rc;
+
+  bufs[0] = uv_buf_init("hello,", sizeof("hello,") - 1);
+  bufs[1] = uv_buf_init(" world", sizeof(" world") - 1);
+
+  do
+    rc = uv_try_write((uv_stream_t*) handle, bufs, ARRAY_SIZE(bufs));
+  while (rc > 0);
+
+  ASSERT_EQ(rc, UV_EAGAIN);
+}
+
+
+static void alloc_cb(uv_handle_t* handle, size_t size, uv_buf_t* buf) {
+  static char base[256];
+
+  buf->base = base;
+  buf->len = sizeof(base);
+}
+
+
+static void read_cb(uv_stream_t* handle, ssize_t nread, const uv_buf_t* buf) {
+  if (spam == spam_0) {
+    ASSERT_UINT64_GT(nread, 0);  /* Expect some bytes. */
+  } else {
+    ASSERT_UINT64_EQ(nread, (ssize_t) buf->len);  /* Expect saturation. */
+  }
+
+  if (handle == (uv_stream_t*) &peer_handle) {
+    spam(&client_handle);
+  } else {
+    uv_close((uv_handle_t*) &peer_handle, NULL);
+    uv_close((uv_handle_t*) &client_handle, NULL);
+    uv_close((uv_handle_t*) &server_handle, NULL);
+  }
```

Comment:
IIUC, I don't see any writes to `client_handle` initiated?
```suggestion
  assert(handle == (uv_stream_t*) &client_handle);
  
  if (spam == spam_0) {
    ASSERT_UINT64_GT(nread, 0);  /* Expect some bytes. */
  } else {
    ASSERT_UINT64_EQ(nread, (ssize_t) buf->len);  /* Expect saturation. */
  }

  uv_close((uv_handle_t*) &peer_handle, NULL);
  uv_close((uv_handle_t*) &client_handle, NULL);
  uv_close((uv_handle_t*) &server_handle, NULL);
```

Reply:
Or maybe you meant:
```suggestion
  if (spam == spam_0) {
    ASSERT_UINT64_GT(nread, 0);  /* Expect some bytes. */
  } else {
    ASSERT_UINT64_EQ(nread, (ssize_t) buf->len);  /* Expect saturation. */
  }

  if (handle == (uv_stream_t*) &client_handle) {
    spam(&client_handle);
  } else {
    uv_close((uv_handle_t*) &peer_handle, NULL);
    uv_close((uv_handle_t*) &client_handle, NULL);
    uv_close((uv_handle_t*) &server_handle, NULL);
  }
```